cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramWrapper.java

```diff
@@ -172,77 +167,19 @@ private void runUserProgram(SparkProgram sparkProgram) {
    * Stops the Spark program by calling {@link org.apache.spark.SparkContext#stop()}
    */
   public static void stopSparkProgram() {
-
-    sparkContextStopBugFixer(); // to close the selector which causes a thread deadlock
-
-    // Now stop the program
-    if (isScalaProgram()) {
-      ((org.apache.spark.SparkContext) getSparkContext().getOriginalSparkContext()).stop();
-    } else {
-      ((org.apache.spark.api.java.JavaSparkContext) getSparkContext().getOriginalSparkContext()).stop();
-    }
-  }
-
-  /**
-   * Fixes the thread deadlock issue in {@link org.apache.spark.SparkContext#stop} where the {@link Selector} field
-   * in {@link ConnectionManager} waits for an interrupt.
-   */
-  private static void sparkContextStopBugFixer() {
-    ConnectionManager connectionManager = getConnectionManager(getSparkContext());
-    if (!closeSelector(connectionManager)) {
-      LOG.warn("Failed to get the Selector which can cause thread deadlock in SparkContext.stop()");
-    }
-  }
-
-  /**
-   * Gets the {@link Selector} field in the {@link ConnectionManager} and closes it which makes it come out of deadlock
-   *
-   * @param connectionManager : the {@link ConnectionManager} of this {@link SparkContext}
-   */
-  private static boolean closeSelector(ConnectionManager connectionManager) {
-    // Get the selector field from the ConnectionManager and make it accessible
-    boolean selectorClosed = false;
-    for (Field field : connectionManager.getClass().getDeclaredFields()) {
-      if (Selector.class.isAssignableFrom(field.getType())) {
-        if (!field.isAccessible()) {
-          field.setAccessible(true);
-        }
-        try {
-          Selector selector = (Selector) field.get(connectionManager);
-          selector.close();
-          selectorClosed = true;
-          break;
-        } catch (IllegalAccessException iae) {
-          LOG.warn("Unable to access the selector field", iae);
-          throw Throwables.propagate(iae);
-        } catch (IOException ioe) {
-          LOG.info("Close on Selector threw IOException", ioe);
-          throw Throwables.propagate(ioe);
-        }
+    if (getSparkContext() != null) {
```

Comment:
Nope. They removed the ConnectionManager class entirely.
I tested the Stop functionality to work without the hack and it works well.